vpn/ciphers.go

```diff
@@ -1,89 +1,152 @@
+	errPadding = errors.New("padding error")
+
+	// errUnsupportedCipher indicates we don't support the desired cipher.
+	errUnsupportedCipher = errors.New("unsupported cipher")
+
+	// errUnsupportedMode indicates that the mode is not uspported.
+	errUnsupportedMode = errors.New("unsupported mode")
 )
 
+// dataCipher encrypts and decrypts OpenVPN data.
 type dataCipher interface {
+	// keySizeBytes returns the key size (in bytes).
 	keySizeBytes() int
+
+	// isAEAD returns whether this cipher has AEAD properties.
 	isAEAD() bool
+
+	// blockSize returns the expected block size.
 	blockSize() int
+
+	// encrypt encripts a plaintext.
+	//
+	// Arguments:
+	//
+	// - key is the key, whose size must be consistent with the cipher;
+	//
+	// - iv is the initialization vector;
+	//
+	// - plaintext is the plaintext to encrypt;
+	//
+	// - ad contains the additional data (optional and only used for AEAD ciphers).
+	//
+	// Returns the ciphertext on success and an error on failure.
 	encrypt(key, iv, plaintext, ad []byte) ([]byte, error)
+
+	// decrypt is the opposite operation of encrypt. It takes in input the
+	// ciphertext and returns the plaintext of an error.
 	decrypt(key, iv, ciphertext, ad []byte) ([]byte, error)
 }
 
-type aesCipher struct {
-	keySizeBits int
-	mode        string
+// dataCipherAES implements dataCipher for AES.
+type dataCipherAES struct {
+	// ksb is the key size in bytes
+	ksb int
+
+	// mode is the cipher mode
+	mode cipherMode
 }
 
-func (a *aesCipher) keySizeBytes() int {
-	return a.keySizeBits / 8
+var _ dataCipher = &dataCipherAES{} // Ensure we implement dataCipher
+
+// keySizeBytes implements dataCipher.keySizeBytes
+func (a *dataCipherAES) keySizeBytes() int {
+	return a.ksb
 }
 
-func (a *aesCipher) isAEAD() bool {
-	if a.mode == cbcMode {
-		return false
-	}
-	return true
+// isAEAD implements dataCipher.isAEAD
+func (a *dataCipherAES) isAEAD() bool {
+	return a.mode != cipherModeCBC
 }
 
-func (a *aesCipher) blockSize() int {
-	if a.mode == cbcMode || a.mode == gcmMode {
+// blockSize implements dataCipher.BlockSize
+func (a *dataCipherAES) blockSize() int {
+	switch a.mode {
+	case cipherModeCBC, cipherModeGCM:
 		return 16
+	default:
+		return 0
 	}
-	return 0
 }
 
-// decrypt tries to decrypt the ciphertext. ad is optional, and only used in AEAD modes.
-func (a *aesCipher) decrypt(key, iv, ciphertext, ad []byte) ([]byte, error) {
-	k := key[:a.keySizeBytes()] // use stdlib
+// decrypt implements dataCipher.decrypt
+func (a *dataCipherAES) decrypt(key, iv, ciphertext, ad []byte) ([]byte, error) {
+	if len(key) != a.keySizeBytes() {
+		return nil, errInvalidKeySize
+	}
 
-	var block cipher.Block
-	block, err := aes.NewCipher(k)
+	block, err := aes.NewCipher(key)
 	if err != nil {
 		return nil, err
 	}
 
-	var mode cipher.BlockMode
-
 	switch a.mode {
-	case cbcMode:
+	case cipherModeCBC:
 		i := iv[:block.BlockSize()]
-		mode = cipher.NewCBCDecrypter(block, i)
+		mode := cipher.NewCBCDecrypter(block, i)
```

Comment:
Isn't it strange that the IV is used more than once?! (Maybe I'm just ignorant 😬)

Reply:
it's not (see data.go). in AEAD mode, they iv is the packetID + hmac[:8]. In CBC/other, it's just random bytes.